Build failure: uasm

## Steps To Reproduce

Steps to reproduce the behavior:

1. build uasm from unstable (`github:nixos/nixpkgs/634fd46801442d760e09493a794c4f15db2d0cbb`).

## Build log

<!-- insert build log in code block in collapsable section -->

<details>
<summary>Build Log</summary>

```
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/w98ia5fj94bmi7zm17wq63cc3rya2hrh-source
source root is source
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
applying patch /nix/store/96jj8d9f05shwww7nrzx60k4q4077p8s-fix-v2_55-compilation-on-macos.patch
patching file ClangOSX64.mak
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
no configure script, doing nothing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
build flags: -j3 SHELL=/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/bash CC=cc
mkdir GccUnixR
cc -D __UNIX__ -c -IH -D __UNIX__ -DNDEBUG -O2 -funsigned-char -Wwrite-strings   -o GccUnixR/main.o main.c
cc -D __UNIX__ -c -IH -D __UNIX__ -DNDEBUG -O2 -funsigned-char -Wwrite-strings   -o GccUnixR/apiemu.o apiemu.c
cc -D __UNIX__ -c -IH -D __UNIX__ -DNDEBUG -O2 -funsigned-char -Wwrite-strings   -o GccUnixR/orgfixup.o orgfixup.c
cc -D __UNIX__ -c -IH -D __UNIX__ -DNDEBUG -O2 -funsigned-char -Wwrite-strings   -o GccUnixR/assemble.o assemble.c
main.c: In function 'main':
main.c:95:17: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
   95 |         argv[0] = pEnv;
      |                 ^
cc -D __UNIX__ -c -IH -D __UNIX__ -DNDEBUG -O2 -funsigned-char -Wwrite-strings   -o GccUnixR/assume.o assume.c
cc -D __UNIX__ -c -IH -D __UNIX__ -DNDEBUG -O2 -funsigned-char -Wwrite-strings   -o GccUnixR/atofloat.o atofloat.c
assemble.c: In function 'PassOneChecks':
assemble.c:1042:32: warning: comparison between pointer and integer
 1042 |         if ( curr->sym.altname > 1 )
      |                                ^
assemble.c: In function 'OnePass':
assemble.c:1181:17: error: implicit declaration of function 'AddSimdTypes' [-Wimplicit-function-declaration]
 1181 |                 AddSimdTypes();
      |                 ^~~~~~~~~~~~
In file included from assemble.c:18:
assemble.c: In function 'get_module_name':
H/globals.h:51:17: error: implicit declaration of function 'strupr'; did you mean '_strupr'? [-Wimplicit-function-declaration]
   51 | #define _strupr strupr
      |                 ^~~~~~
assemble.c:1280:5: note: in expansion of macro '_strupr'
 1280 |     _strupr( ModuleInfo.name );
      |     ^~~~~~~
assemble.c: In function 'GetExt':
assemble.c:1430:23: warning: return discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
 1430 |                 return( EXE_EXT );
      |                       ^
assemble.c:1433:23: warning: return discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
 1433 |                 return( BIN_EXT );
      |                       ^
assemble.c:1435:15: warning: return discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
 1435 |         return( OBJ_EXT );
      |               ^
assemble.c:1437:15: warning: return discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
 1437 |         return( LST_EXT );
      |               ^
assemble.c:1439:15: warning: return discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
 1439 |         return( ERR_EXT );
      |               ^
In file included from assemble.c:94:
assemble.c: In function 'AssembleModule':
H/Colors.h:12:15: warning: format '%lu' expects argument of type 'long unsigned int', but argument 3 has type 'uint_32' {aka 'unsi>
   12 | #define KWHT  "\x1B[37m"
      |               ^~~~~~~~~~
H/Colors.h:20:17: note: in expansion of macro 'KWHT'
   20 | #define FWHT(x) KWHT x RST
      |                 ^~~~
assemble.c:1801:56: note: in expansion of macro 'FWHT'
 1801 |                                                 printf(FWHT("%s: %lu lines, "), GetFNamePart(GetFName(ModuleInfo.srcfile)->
      |                                                        ^~~~
make: *** [gccLinux64.mak:35: GccUnixR/assemble.o] Error 1
make: *** Waiting for unfinished jobs....
```

</details>

## Additional context

From what I can see, probably related to new gcc 14 default.
I set `nixpkgs.hostPlatform.gcc.{arch,tune}` to `"znver4"`, which may be relevant.

The package was being built as a dependency of `7zz`, because I had had `useUasm` enabled because the feature flag didn't work correctly before a recent commit.
I have since disabled it, and `7zz` builds fine, so this is no longer relevant to me, but still a failure that should probably be addressed.

<!-- Add any other context about the problem here. -->

## Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->
 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.5-xanmod1, NixOS, 25.05 (Warbler), 25.05.20241219.d70bd19`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.91.1
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux, x86_64-v4-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/alfarel/.config/nix/nix.conf:/nix/store/v5pkh7wda19iw13vsc4ma7cr8sbhzxs1-stylix-kde-config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/alfarel/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/alfarel/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/alfarel/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/3vdv3p0nhfzz4k3xv5dfmn6qw653xc2z-lix-2.91.1/share`
 - nixpkgs: `/nix/store/8fwsiv0hd7nw1brkvka0jf1frk3m7qkr-source`


## Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
@thiagokokada 

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/ua/uasm/package.nix
+++ b/pkgs/by-name/ua/uasm/package.nix
@@ -6,7 +6,10 @@
   uasm,
 }:
 
-gcc13Stdenv.mkDerivation rec {
+let
+  stdenv = gcc13Stdenv;
+in
+stdenv.mkDerivation rec {
   pname = "uasm";
   version = "2.57";
 
@@ -25,7 +28,7 @@ gcc13Stdenv.mkDerivation rec {
     else
       "Makefile-Linux-GCC-64.mak";
 
-  makeFlags = [ "CC=${gcc13Stdenv.cc.targetPrefix}cc" ];
+  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
 
   installPhase = ''
     runHook preInstall
@@ -49,5 +52,6 @@ gcc13Stdenv.mkDerivation rec {
     platforms = platforms.unix;
     maintainers = with maintainers; [ thiagokokada ];
     license = licenses.watcom;
+    broken = stdenv.isDarwin;
   };
 }

--- a/pkgs/by-name/ua/uasm/package.nix
+++ b/pkgs/by-name/ua/uasm/package.nix
@@ -1,13 +1,13 @@
 {
   lib,
-  stdenv,
+  gcc13Stdenv,
   fetchFromGitHub,
   fetchpatch,
   testers,
   uasm,
 }:
 
-stdenv.mkDerivation rec {
+gcc13Stdenv.mkDerivation rec {
   pname = "uasm";
   version = "2.56.2";
 
@@ -31,9 +31,9 @@ stdenv.mkDerivation rec {
 
   enableParallelBuilding = true;
 
-  makefile = if stdenv.hostPlatform.isDarwin then "ClangOSX64.mak" else "gccLinux64.mak";
+  makefile = if gcc13Stdenv.hostPlatform.isDarwin then "ClangOSX64.mak" else "gccLinux64.mak";
 
-  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
+  makeFlags = [ "CC=${gcc13Stdenv.cc.targetPrefix}cc" ];
 
   installPhase = ''
     runHook preInstall

--- a/pkgs/by-name/ua/uasm/package.nix
+++ b/pkgs/by-name/ua/uasm/package.nix
@@ -2,36 +2,28 @@
   lib,
   gcc13Stdenv,
   fetchFromGitHub,
-  fetchpatch,
   testers,
   uasm,
 }:
 
 gcc13Stdenv.mkDerivation rec {
   pname = "uasm";
-  version = "2.56.2";
+  version = "2.57";
 
   src = fetchFromGitHub {
     owner = "Terraspace";
-    repo = pname;
-    # Specifying only the tag results in the following error during download:
-    # the given path has multiple possibilities: #<Git::Ref:0x00007f618689c378>, #<Git::Ref:0x00007f618689c1e8>
-    # Probably because upstream has both a tag and a branch with the same name
-    rev = "refs/tags/v${version}";
-    hash = "sha256-QiRBscY6zefeLDDVhS/+j9yIJ+5QhgkDQh1CLl/CslM=";
+    repo = "uasm";
+    tag = "v${version}r";
+    hash = "sha256-HaiK2ogE71zwgfhWL7fesMrNZYnh8TV/kE3ZIS0l85w=";
   };
-
-  patches = [
-    (fetchpatch {
-      name = "fix-v2_55-compilation-on-macos.patch";
-      url = "https://github.com/Terraspace/UASM/commit/b50c430cc3083c7f32e288a9f64fe1cafb03091d.patch";
-      sha256 = "sha256-FGFB282LSEKtGD1cIRH+Qi5bye5Gx4xb0Ty4J03xjCU";
-    })
-  ];
 
   enableParallelBuilding = true;
 
-  makefile = if gcc13Stdenv.hostPlatform.isDarwin then "ClangOSX64.mak" else "gccLinux64.mak";
+  makefile =
+    if gcc13Stdenv.hostPlatform.isDarwin then
+      "Makefile-OSX-Clang-64.mak"
+    else
+      "Makefile-Linux-GCC-64.mak";
 
   makeFlags = [ "CC=${gcc13Stdenv.cc.targetPrefix}cc" ];
 

